docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,84 @@
+# OBJECT LEVEL SECURITY
+
+EVE software receives its configuration from EVC. The configuration may have confidential portions, such as passwords, acces-keys etc. Even though the configuration is exchanged over a secure channel(TLS), in case of HTTPS proxies with lawful inspection, there is a chance that these sensitive information may be exposed as plain-text for a brief period of time. In the interest of protecting these sensitive objects from getting exposed during transit, we need to deploy an additional mechanism on top of TLS to protect sensitive sections of the configuration. Using object level security, one can encrypt these objects at the origin, and by using zero-knowledge key sharing mechanism, such as ECDH, one can share the secret key used for decryption at the receiving end.  Thus a proxy, without access to this secret key, will not be able to see the actual credentials/secrets etc. 
+
```

Comment:
I think it is extremely important that, after the above rationale ("why are we doing this") section, we need a high-level architecture ("how are we doing this") section, before we get into details of message types and algorithms and such.

It has to include:

* what are the certs? How many?
* who signs them?
* how does this protect against man in the middle when it is a legitimate mitm?
* show the flow of startup, very similar to how TLS setup is shown, and especially mTLS. See for example cloudflare's tls description [here](https://www.cloudflare.com/learning/ssl/what-happens-in-a-tls-handshake/) or mtls description [here](https://developers.cloudflare.com/access/service-auth/mtls/)
* how do we ensure perfect forward secrecy?
* very precisely, how does the mitm get stymied when it attempts to read the encrypted+signed message but not the body itself which is encrypted with server-side TLS?
* do we support mTLS and fallback to this model? If not, why; if so, how?

Reply:
@deitch do you want explicit answers to the above? I'll update the PR and you can check if you see the answers.

Note that this is layered on top of 1) the transport security between the TLS endpoints and the 2) object signing providing integrity and sender authentication between the microservices. It depends on the integrity and sender authentication to prevent man in the middle attacks.

(P)FS is provided by having the controller periodically generate a new ECC certificate.

Note that ECDH just like DH doesn't have any message exchanges like TLS; it is just a generation of a shared secret using two public/private key pairs.

Reply:
I think it answered most of them. I was able to "reverse engineer" the flow from the text, but it would be great if we could show even an ascii diagram of flow of secret from controller through encryption over the internet to zedagent to specific microservice decrypting and using it.

Reply:
There is no secret flowing. There are controller certs and IVs flowing. But I'm not the person to draw the diagram. I'd prefer to ship this and iterate as we complete the implementation (for non-TPM and controller cert refresh)

Reply:
Sure there are secrets flowing. For example, the controller contains a WiFi password that should be available only to nim. It has to get from the controller to nim via the various channels, being encrypted along the way.

But I get what you are saying about "good enough for now and improve later."